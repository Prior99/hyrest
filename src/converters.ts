import { FullValidator, Processed } from "./validation";

export interface Converted<T> {
    error?: string;
    value?: T;
}

export type Converter<T> = (input: any) => Converted<T> | Promise<Converted<T>> | Processed<T> | Promise<Processed<T>>;

/**
 * Converts the given input to an integer if possible.
 *
 * @param input The input to convert to an integer. Can be anything in any form.
 *
 * @return An integer if the conversion succeeded or an error if the input was not a valid
 *         integer.
 */
export function int(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseInt(input);
    if (isNaN(value) || parseFloat(input) !== value) { return { error: "Not a valid integer." }; }
    return { value };
}

/**
 * Converts the given input to a float if possible.
 *
 * @param input The input to convert to a float. Can be anything in any form.
 *
 * @return A float if the conversion succeeded or an error if the input was not a valid
 *         float.
 */
export function float(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseFloat(input);
    if (isNaN(value)) { return { error: "Not a valid float." }; }
    return { value };
}

/**
 * Makes sure the given input is a string.
 *
 * @param input The input to check.
 *
 * @return The string in `input` if it was a string and an error otherwise.
 */
export function str(value: any): Converted<string> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "string") { return { error: "Not a valid string." }; }
    return { value };
}

/**
 * Makes sure the given input is an object.
 *
 * @param input The input to check.
 *
 * @return The input if it was an object and an error otherwise.
 */
export function obj(value: any): Converted<Object> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "object" || Array.isArray(value)) { return { error: "Not a valid object."}; }
    return { value };
}

/**
 * Makes sure the given input is a boolean.
 *
 * @param input The input to check.
 *
 * @return The boolean in `input` if it was a boolean and an error otherwise.
 */
export function bool(value: any): Converted<boolean> {
    if (typeof value === "undefined") { return { value }; }
    if (value === "true" || value === "false") { return { value: value === "true" }; }
    if (typeof value !== "boolean") { return { error: "Not a valid boolean." }; }
    return { value };
}
