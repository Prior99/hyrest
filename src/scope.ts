import "reflect-metadata";

import { Constructable } from "./types";
import * as invariant from "invariant";
import { allKeys } from "./all-keys";
import { getTransforms } from "./transform";
import { isCustomClass } from "./validation";

export type TypeCreator = () => Constructable<any>;

export interface SpecifiedTypes {
    /**
     * Types specified for method's parameters.
     */
    params: Map<number, TypeCreator>;
    /**
     * The type specified for the property itself.
     */
    property?: TypeCreator;
}

export interface PropertyMeta {
    /**
     * The prototype of the class on which the property was decorated.
     */
    readonly target: Object;
    /**
     * The name of the property that was decorated.
     */
    readonly property: string;
    /**
     * The expected type for this property.
     */
    readonly expectedType: any | any[];
}

export class Scope {
    private included: Scope[] = [];
    private ownProperties: PropertyMeta[] = [];

    /**
     * Include all properties from another scope in this scope:
     *
     * **Example:**
     *
     * ```
     * const exampleA = createScope();
     * const exampleB = createScope().include(exampleA);
     *
     * class A {
     *     @scope(exampleA)
     *     public property1: string;
     *
     *     @scope(exampleB)
     *     public property2: string;
     * }
     *
     * console.log(exampleA.properties); // Only `property1`.
     * console.log(exampleB.properties); // Will be `property1` and `property2`.
     * ```
     * @param scopes A list of scopes to include in this scope.
     *
     * @return This instance to use this method for the builder pattern.
     */
    public include(...scopes: Scope[]): Scope {
        scopes.forEach(includedScope => this.included.push(includedScope));
        // Builder pattern: Return `this`.
        return this;
    }

    public registerProperty(property: PropertyMeta) {
        this.ownProperties.push(property);
    }

    /**
     * All properties decorated by `@scope` with this scope and all properties
     * included by using `include()`.
     */
    public get properties() {
        return this.included.reduce((result, included) => {
            result.push(...included.properties);
            return result;
        }, [ ...this.ownProperties ]);
    }

    /**
     * Will return all properties for a specific class.
     *
     * **Example:**
     *
     * ```
     * const example = createScope();
     *
     * class A {
     *     @scope(example)
     *     public property1: string;
     * }
     *
     * class B {
     *     @scope(example)
     *     public property2: string;
     * }
     *
     * console.log(example.propertiesForClass(A)); // Only `property1` is returned.
     * console.log(example.propertiesForClass(A)); // Only `property2` is returned.
     * ```
     *
     * This method takes all properties included using `include()` into account.
     *
     * @param clazz The clazz to return the properties for.
     *
     * @return All properties from the specified class.
     */
    public propertiesForClass(clazz: Function) {
        return this.properties.filter(property => {
            let current = clazz;
            while (current !== null && typeof current !== "undefined") { // tslint:disable-line
                if (property.target.constructor === current) {
                    return true;
                }
                current = Object.getPrototypeOf(current);
            }
        });
    }
}

/**
 * Create a new scope.
 *
 * @return A new scope.
 */
export function createScope() {
    return new Scope();
}

/**
 * Use this decorator to mark a property, getter or a setter on a class as belonging to a scope
 * so they are included when `dump()`ing or `populate()`ing them.
 *
 * @param scopes The list of scopes in which the decorated property should be included.
 *
 * @return A decorator which can be used to decorate getters, setters and properties.
 *
 * @see dump
 * @see populate
 */
export function scope(...scopes: Scope[]) {
    return function<T>(target: Object, property: string, descriptor?: PropertyDescriptor) {
        const expectedType = Reflect.getMetadata("design:type", target, property);
        scopes.forEach(decoratedScope => decoratedScope.registerProperty({
            target, property, expectedType,
        }));
    };
}

/**
 * Returns the specified types for a property and its parameters. Is always guaranteed to
 * return an object. If non existed, a new one is created.
 *
 * @param target The target object of which to get the specified types.
 * @param property The property on `target` on which to get the specified types.
 *
 * @return An object containing the type of the property and a map for all its parameters.
 */
export function getSpecifiedType(target: Object, property: string | symbol): SpecifiedTypes {
    let specified = Reflect.getMetadata("specifytype", target, property);
    if (typeof specified === "undefined") {
        specified = {
            params: new Map(),
        };
        Reflect.defineMetadata("specifytype", specified, target, property);
    }
    return specified;
}

/**
 * Help to specify the type of a property or paramter. Might be necessary for arrays and cyclic
 * depdendencies.
 *
 * @param factory A function with no arguments returning the desired type.
 *
 * @return A decorator for a property or a parameter.
 */
export function specify<T>(factory: TypeCreator) {
    return function(target: Object, property: string | symbol, arg3?: any) {
        const specified = getSpecifiedType(target, property);
        if (typeof arg3 === "number") {
            specified.params.set(arg3, factory);
            return;
        }
        specified.property = factory;
    };
}

export function dump<T>(dumpScope: Scope): (instance: T) => T;
export function dump<T>(dumpScope: Scope, instance: T): T;
/**
 * Dump a structure with a given scope. This can either be invoked using the curried syntax
 * (`dump(scope)(obj)` or regularly (`dump(scope, obj)`).
 * This will dump the structure as new, plain objects, including only the keys marked with the
 * `@scope` decorator and the specified scope.
 *
 * @param dumpScope The scope used for dumping.
 * @param arg2 The instance to dump.
 *
 * @return The dumped, stripped down version of the input object.
 */
export function dump<T>(dumpScope: Scope, arg2?: T): T | ((instance: T) => T) {
    function internalDump<U extends any | any[]>(instance: U): U {
        if (Array.isArray(instance)) {
            return instance.map(internalDump) as any as U;
        }
        if (typeof instance === "object" && instance !== null) {
            const propertiesForClass = dumpScope.propertiesForClass(instance.constructor);
            const keys = allKeys(instance);
            return keys.reduce((result, key) => {
                if (propertiesForClass.find(({ property }) => property === key)) {
                    (result as any)[key] = internalDump((instance as any)[key]);
                }
                return result;
            }, {}) as U;
        }
        return instance;
    }
    if (typeof arg2 !== "undefined") {
        return internalDump(arg2);
    }
    return internalDump;
}

export function populate<T>(populateScope: Scope, initialClass: Constructable<T>): (data: any) => T;
export function populate<T>(populateScope: Scope, initialClass: Constructable<T>, data: any): T;
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arrayType: Constructable<T>): (data: any) => T;
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arrayType: Constructable<T>, data: any): T;
/**
 * Populate a structure of classes using an object and a scope.
 * This will recursively create instances of all classes and populate them with the nested objects in
 * the input.
 *
 * @param populateScope The scope to populate. Only properties included in this scope will be populated.
 * @param initialClass The class to start populating with.
 *             It can also be invoked with an additional array type.
 * @param arg4 When populating an array, an additional array type must be specified as the thrid argument
 *             and the data is specified as the fourth argument instead.
 *
 * @return A curried function for use in higher order functions if `arg3` is not specified and the populated
 *         structure otherwise.
 */
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arg3?: any, arg4?: any,
): T | ((data: T) => T) {
    function internalPopulate<U extends any | any[]>(
        data: any, thisClass: any, specifiedClass: Constructable<any>,
    ): U {
        // Perform population of an array.
        if (thisClass === Array) {
            invariant(Array.isArray(data), "Structure does not match. Array expected.");
            invariant(typeof specifiedClass === "function", "Array type not specified.");
            return (data as any[]).map(element => internalPopulate(element, specifiedClass, undefined)) as any as U;
        }
        const guardedClass = thisClass || specifiedClass;
        // Ignore primitives.
        if (guardedClass === Number || guardedClass === Boolean || guardedClass === String || guardedClass === Object) {
            return data;
        }
        invariant(typeof data === "object" && !Array.isArray(data), "Structure does not match. Object expected.");
        if (typeof guardedClass !== "function") {
            throw new Error("Could not infer type. This might be due to a cyclic dependency.");
        }
        // Instanciate the classe to populate.
        const instance = new guardedClass();
        // Get a list of all properties to populate.
        const propertiesForClass = populateScope.propertiesForClass(guardedClass);
        propertiesForClass.forEach(({ property, target, expectedType }) => {
            const dataValue = (data as any)[property];
            if (typeof dataValue === "undefined") {
                return;
            }
            const specifyTypeCreator = getSpecifiedType(target, property).property;
            const specifyType = specifyTypeCreator && specifyTypeCreator();
            const transforms = getTransforms(target, property);
            const populated = internalPopulate(dataValue, expectedType, specifyType);
            if (transforms.propertyTransform) {
                (instance as any)[property] = transforms.propertyTransform(populated);
            } else {
                (instance as any)[property] = populated;
            }
        });
        return instance;
    }
    // `arg3` can either be array type or data.
    if (typeof arg3 !== "undefined") {
        // `arg3 is the array type and `arg4` is the data.
        if (typeof arg4 !== "undefined") {
            return internalPopulate(arg4, initialClass, arg3);
        }
        // If `arg3` is a `Constructable`: It's the array type and a curried version shall be returned.
        if (typeof arg3 === "function") {
            return (data: any) => internalPopulate(data, initialClass, arg3);
        }
        // `arg3 is simply the data and no array type is specified.
        return internalPopulate(arg3, initialClass, undefined);
    }
    return (data: any) => internalPopulate(data, initialClass, undefined);
}
