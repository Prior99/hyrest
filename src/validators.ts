export interface Validation {
    error?: string;
}

export type Validator<T> = (input: T) => Validation | Promise<Validation>;

/**
 * Enforces `input` to be one of the values specified in `options`.
 *
 * @param options The options of which `input` needs to be.
 *
 * @return A Converter which checks if the input is one of `options` and returns an error if that was
 *         not the case and the `input` itself otherwise.
 */
export function oneOf<T>(...options: T[]): Validator<T> {
    return (value: T) => {
        if (typeof value === "undefined") { return {}; }
        if (!options.includes(value)) { return { error: `Not one of (${options.join(", ")}).` }; }
        return {};
    };
}

/**
 * Makes sure the given input is not `null` or `undefined`.
 *
 * @param input The input to check.
 *
 * @return The input if it was not `undefined` or `null`, otherwise an error will be returned.
 */
export function required<T>(value: T): Validation {
    if (typeof value === "undefined" || value === null) { return { error: "Missing required field." }; }
    return {};
}
