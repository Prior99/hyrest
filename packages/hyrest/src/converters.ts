import { FullValidator } from "./validation";
import { Processed } from "./processed";
import { Scope } from "./scope";

export interface Converted<T> {
    error?: string;
    value?: T;
}

export type Converter<T> = (input: any, scope?: Scope, context?: any) =>
    Converted<T> |
    Promise<Converted<T>> |
    Processed<T> |
    Promise<Processed<T>>;

/**
 * Converts the given input to an integer if possible.
 *
 * @param input The input to convert to an integer. Can be anything in any form.
 *
 * @return An integer if the conversion succeeded or an error if the input was not a valid
 *         integer.
 */
export function int(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseInt(input);
    if (isNaN(value) || parseFloat(input) !== value) { return { error: "Not a valid integer." }; }
    return { value };
}

/**
 * Converts the given input to a float if possible.
 *
 * @param input The input to convert to a float. Can be anything in any form.
 *
 * @return A float if the conversion succeeded or an error if the input was not a valid
 *         float.
 */
export function float(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseFloat(input);
    if (isNaN(value)) { return { error: "Not a valid float." }; }
    return { value };
}

/**
 * Makes sure the given input is a string.
 *
 * @param input The input to check.
 *
 * @return The string in `input` if it was a string and an error otherwise.
 */
export function str(value: any): Converted<string> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "string") { return { error: "Not a valid string." }; }
    return { value };
}

/**
 * Makes sure the given input is an object.
 *
 * @param input The input to check.
 *
 * @return The input if it was an object and an error otherwise.
 */
export function obj(value: any): Converted<Object> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "object" || Array.isArray(value)) { return { error: "Not a valid object."}; }
    return { value };
}

/**
 * Makes sure the given input is a boolean.
 *
 * @param input The input to check.
 *
 * @return The boolean in `input` if it was a boolean and an error otherwise.
 */
export function bool(value: any): Converted<boolean> {
    if (typeof value === "undefined") { return { value }; }
    if (value === "true" || value === "false") { return { value: value === "true" }; }
    if (typeof value !== "boolean") { return { error: "Not a valid boolean." }; }
    return { value };
}

/**
 * Makes sure the given input is an array with all elements matching the given criteria.
 *
 * @param validator A validator to apply to all element of the array.
 *
 * @return The array converter.
 */
export function arr<T, TContext>(validator?: FullValidator<T, TContext>): Converter<T[]> {
    return async (value: any, scope?: Scope, context?: any) => {
        const processed = new Processed<T[]>();
        // Ignore `undefined` inputs. This is handled by the `required` validator if intended by
        // the user.
        if (typeof value === "undefined") { return processed; }
        // If the input is not an array at all, don't even try to validate any elements.
        if (!Array.isArray(value)) {
            processed.addErrors("Not an array.");
            return processed;
        }
        // Apply the validator (if provided) to all elements and store the results in the `nested`
        // property of the result. The keys will be the array indices.
        if (typeof validator !== "undefined"){
            const results = await Promise.all(value.map(async (elem, index) => {
                const result = await validator(elem, { scope, context });
                if (result.hasErrors) {
                    processed.addNested(index, result);
                }
                return result;
            }));
            if (!results.find(result => result.hasErrors)) {
                processed.value = results.map(result => result.value);
            }
            return processed;
        }
        processed.value = value;
        return processed;
    };
}

/**
 * Makes sure the given input is `null`.
 *
 * @param input The input to check.
 *
 * @return `null` if it was a `null` and an error otherwise.
 */
export function empty(value: any): Converted<null> {
    if (value === null) { return { value }; }
    return { error: "Not null." };
}
