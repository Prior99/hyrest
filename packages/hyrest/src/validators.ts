import { Scope } from "./scope";

export interface Validation {
    error?: string;
}

export type Validator<T> = (input: T, scope?: Scope) => Validation | Promise<Validation>;

/**
 * Enforces `input` to be one of the values specified in `options`.
 *
 * @param options The options of which `input` needs to be.
 *
 * @return A validator which checks if the input is one of `options` and returns an error if that was
 *         not the case.
 */
export function oneOf<T>(...options: T[]): Validator<T> {
    return (value: T) => {
        if (typeof value === "undefined") { return {}; }
        if (!options.includes(value)) { return { error: `Not one of (${options.join(", ")}).` }; }
        return {};
    };
}

/**
 * Makes sure the given input is not `null` or `undefined`.
 *
 * @param input The input to check.
 *
 * @return The input if it was not `undefined` or `null`, otherwise an error will be returned.
 */
export function required<T>(value: T): Validation {
    if (typeof value === "undefined" || value === null) { return { error: "Missing required field." }; }
    return {};
}

export function length({ min, max }: { max: number, min: number }): Validator<string | any[]>;
export function length(min: number, max: number): Validator<string | any[]>;
/**
 * Validates the input string to be at least of length `min` and not longer than `max`.
 *
 * @param options An object with the keys `min` and `max` for the minimum and maximum length of the string.
 *
 * @return A validator which checks the length of the input string.
 */
export function length(arg1: { max: number, min: number } | number, arg2?: number): Validator<string | any[]> {
    const { min, max } = typeof arg1 === "object" ? arg1 : { min: arg1, max: arg2 };
    return (value: string | any[]) => {
        if (typeof value === "undefined") { return {}; }
        if (value === null) { return { error: "Cannot determine length of null." }; }
        if (typeof max !== "undefined" && value.length > max) {
            return { error: `Exceeds maximum length of ${max}.` };
        }
        if (typeof min !== "undefined" && value.length < min) {
            return { error: `Shorter than minimum length of ${min}.` };
        }
        return {};
    };
}

// Taken from http://emailregex.com/
const emailValidationRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/; // tslint:disable-line
// Taken from https://github.com/mixer/uuid-validate
const uuidValidationRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-4][0-9a-f]{3}-[0-9a-f]{4}-[0-9a-f]{12}$/;

/**
 * Validates that the input is an email.
 *
 * @param value The value to check.
 *
 * @return An error if `value` was not an email.
 */
export function email(value: string): Validation {
    if (typeof value === "undefined") { return {}; }
    if (typeof value !== "string" || !value.match(emailValidationRegex)) { // tslint:disable-line
        return { error: `String is not a valid email.` };
    }
    return {};
}

/**
 * Limits a validator to a given scope. The validator will only be executed if the scope matches.
 *
 * @param onlyScope The scope to limit the validator to.
 * @param validator The validator to execute if the scope matches.
 *
 * @return A validator only executed if the scope matches.
 */
export function only<T>(onlyScope: Scope, validator: Validator<T>): Validator<T> {
    return (value: T, currentScope: Scope) => {
        if (currentScope === onlyScope) {
            return validator(value);
        }
        return {};
    };
}

export function range({ min, max }: { max: number, min: number }): Validator<number>;
export function range(min: number, max: number): Validator<number>;
/**
 * Validates the input to be a number in the specified range.
 *
 * @param options An object with the keys `min` and `max` for the minimum and maximum length of the string.
 *
 * @return A validator which checks the range of the input string.
 */
export function range(arg1: { max: number, min: number } | number, arg2?: number): Validator<number> {
    const { min, max } = typeof arg1 === "object" ? arg1 : { min: arg1, max: arg2 };
    return (value: number) => {
        if (typeof value === "undefined") { return {}; }
        if (value === null) { return { error: "Cannot compare with null." }; }
        if (typeof max !== "undefined" && value > max) {
            return { error: `Exceeds maximum of ${max}.` };
        }
        if (typeof min !== "undefined" && value < min) {
            return { error: `Less than minimum of ${min}.` };
        }
        return {};
    };
}

/**
 * Validates that the input is an uuid.
 *
 * @param value The value to check.
 *
 * @return An error if `value` was not an uuid.
 */
export function uuid(value: string): Validation {
    if (typeof value === "undefined") { return {}; }
    if (typeof value !== "string" || !value.match(uuidValidationRegex)) {
        return { error: `String is not a valid uuid.` };
    }
    return {};
}
