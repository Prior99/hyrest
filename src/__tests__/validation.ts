import { is, getPropertyValidation, validateSchema, processValue, inferConverter } from "../validation";
import { int, float, str, obj, bool, arr } from "../converters";
import { oneOf, required, length } from "../validators";
import { specify } from "../scope";

test("@is as parameter decorator infers the converter", async () => {
    class TestController {
        public method(@is() parameter1: string) {
            return;
        }
    }

    const controller = new TestController();
    const metadata = Reflect.getMetadata("validation:parameters", controller, "method");
    expect(await metadata.get(0).fullValidator("test", {})).toMatchSnapshot();
});

test("@is as parameter decorator", () => {
    const is1 = is(int);
    const is2 = is(float);
    const is3 = is(str).validate(oneOf("a", "b"));

    class TestController {
        public method(@is1 parameter1: number, @is2 parameter2: number, @is3 parameter3: string) {
            return;
        }
    }

    const controller = new TestController();

    const metadata = Reflect.getMetadata("validation:parameters", controller, "method");
    const expected = new Map();
    expected.set(0, { fullValidator: is1 });
    expected.set(1, { fullValidator: is2 });
    expected.set(2, { fullValidator: is3 });
    expect(metadata).toEqual(expected);
});

test("@is as property decorator", () => {
    const is1 = is(str).validate(oneOf("a", "b"), required);
    const is2 = is(int).validate(oneOf(1, 2, 3), required);

    class TestController { // tslint:disable-line
        @is1 public test1: string;
        @is2 public test2: number;
    }

    const controller = new TestController();

    const metadata1 = Reflect.getMetadata("validation:property", controller, "test1");
    expect(metadata1).toEqual({ fullValidator: is1 });
    const metadata2 = Reflect.getMetadata("validation:property", controller, "test2");
    expect(metadata2).toEqual({ fullValidator: is2 });
    expect(getPropertyValidation(controller, "test2")).toEqual(metadata2);
});

[
    {
        testSchema: {
            a: is(int),
            b: is(float),
            c: is(str),
        },
        valid: [
            {
                a: 10,
                b: 23.5,
                c: "test",
            },
            {
                a: 10120,
                b: 123.5,
                c: "another string",
            },
        ],
        invalid: [
            {
                a: false,
                b: 123.5,
                c: "another string",
                d: "invalid",
            },
            {
                a: false,
            },
        ],
    },
    {
        testSchema: {
            arraOfArrays: is(arr(
                is(arr(
                    is(int).validate(oneOf(5, 6, 7)),
                )),
            )),
        },
        valid: [
            {
                arraOfArrays: [
                    [5, 6, 7],
                    [6, 7],
                    [],
                ],
            },
        ],
        invalid: [
            {
                arraOfArrays: [5],
            },
            {
                arraOfArrays: [
                    [8],
                ],
            },
        ],
    },
    {
        testSchema: {
            a: is(int).validate(oneOf(1, 2, 3, 4)),
            b: {
                c: is(int),
            },
            d: is(int).validate(required),
            e: is(arr(is(int).validate(oneOf(1, 2, 3)))),
            f: is(arr(is(obj).schema({
                g: is(str).validate(length(1, 5)),
            }))),
        },
        valid: [
            {
                a: 3,
                b: {
                    c: 19,
                },
                d: 8,
            },
            {
                d: 8,
                e: [1, 2, 1, 2, 1],
                f: [{ g: "foo" }, { g: "bar" }, { g: "bas" }, { g: "baz" } ],
            },
        ],
        invalid: [
            {
                a: 19,
                b: {
                    c: {
                        d: "test",
                    },
                },
            },
            {
                d: 8,
                e: [1, 2, 1, 3],
                f: [{ g: "loremipsum" }],
            },
            {
                d: 8,
                e: [5, 6, 20],
                f: [{ g: "test" }],
            },
            undefined,
        ],
    },

].forEach(({ testSchema, valid, invalid }) => {
    valid.forEach(input => {
        test("The test schema detects a valid input as valid", async () => {
            expect(await validateSchema(testSchema, input)).toEqual({});
        });
    });
    invalid.forEach(input => {
        test("The test schema detects a invalid input as invalid", async () => {
            expect(await validateSchema(testSchema, input)).toMatchSnapshot();
            expect((await validateSchema(testSchema, input)).hasErrors).toBe(true);
        });
    });
});

[
    {
        converter: str,
        validators: [oneOf("a", "b")],
        inputs: ["a", "b", "c", "", 1, null, undefined, {}, []], // tslint:disable-line
    },
    {
        converter: int,
        validators: [required],
        inputs: ["c", "", 10, null, undefined, {}, []], // tslint:disable-line
    },
].forEach(testCase => {
    testCase.inputs.forEach(input => {
        test("`processValue` processes inputs as expected", async () => {
            expect(await processValue(input, testCase.converter, testCase.validators)).toMatchSnapshot();
        });
    });
});

[
    {
        validator: arr(),
        tests: [
            [],
            [1, 2, 3],
            {},
            1,
            "",
            "1",
            null, // tslint:disable-line
            undefined,
            true,
        ],
    },
    {
        validator: arr(is(str)),
        tests: [
            [1, 2, 3, 4],
            ["a", "b", "c"],
            [],
            {},
            12,
        ],
    },
    {
        validator: arr(is(int).validate(oneOf(1, 2, 3))),
        tests: [
            [1, 2, 3],
            [1, 2],
            [],
            {},
            12,
        ],
    },
].forEach(testCase => {
    testCase.tests.forEach(input => {
        test("`arr` works as expected", async () => {
            expect(await testCase.validator(input)).toMatchSnapshot();
        });
    });
});

test("`inferConverter`", () => {
    expect(inferConverter(Number)).toBe(float);
    expect(inferConverter(String)).toBe(str);
    expect(inferConverter(Boolean)).toBe(bool);
    expect(inferConverter(Object)).toBe(obj);
    expect(inferConverter(Function)).toBe(obj);

    const stringArrayConverter = inferConverter(Array, String);
    const workingStringArrayConverter = arr(is(str));
    const correctArray = ["a", "b"];
    const incorrectArray = [1, 2];
    expect(stringArrayConverter(correctArray)).toEqual(workingStringArrayConverter(correctArray));
    expect(stringArrayConverter(incorrectArray)).toEqual(workingStringArrayConverter(incorrectArray));

    const workingUnknownArrayConverter = arr();
    const unknownArrayConverter = inferConverter(Array);
    expect(unknownArrayConverter(correctArray)).toEqual(workingUnknownArrayConverter(correctArray));
    expect(unknownArrayConverter(incorrectArray)).toEqual(workingUnknownArrayConverter(incorrectArray));
});

test("`processValue` without a converter", async () => {
    expect(await processValue("test", undefined, [])).toMatchSnapshot();
});

test("`@is` with no chance to infer the type", () => {
    expect(() => is()("test", { context: {} })).toThrowErrorMatchingSnapshot();
});

test("`@is` with the type specified explicitly in a parameter", async () => {
    class B { // tslint:disable-line
        @is()
        public test: string;
    }
    class A { // tslint:disable-line
        public method(@is() @specify(() => B) param: undefined) {
        }
    }
    const a = new A();
    const metadata = Reflect.getMetadata("validation:parameters", a, "method");
    expect(await metadata.get(0).fullValidator({ test: 9 }, {})).toMatchSnapshot();
});
