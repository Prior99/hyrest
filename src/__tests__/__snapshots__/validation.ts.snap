// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@is as parameter decorator 1`] = `
Map {
  2 => Object {
    "converter": [Function],
    "validationSchema": undefined,
    "validatorFactory": undefined,
    "validators": Array [
      [Function],
    ],
  },
  1 => Object {
    "converter": [Function],
    "validationSchema": undefined,
    "validatorFactory": undefined,
    "validators": Array [],
  },
  0 => Object {
    "converter": [Function],
    "validationSchema": undefined,
    "validatorFactory": undefined,
    "validators": Array [],
  },
}
`;

exports[`@is as property decorator 1`] = `
Object {
  "converter": [Function],
  "validationSchema": undefined,
  "validatorFactory": undefined,
  "validators": Array [
    [Function],
    [Function],
  ],
}
`;

exports[`@is as property decorator 2`] = `
Object {
  "converter": [Function],
  "validationSchema": undefined,
  "validatorFactory": undefined,
  "validators": Array [
    [Function],
    [Function],
  ],
}
`;

exports[`The test schema detects a invalid input as invalid 1`] = `
Processed {
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "Not a valid integer.",
      ],
    },
    "d": Processed {
      "errors": Array [
        "Unexpected key.",
      ],
    },
  },
}
`;

exports[`The test schema detects a invalid input as invalid 2`] = `
Processed {
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "Not a valid integer.",
      ],
    },
  },
}
`;

exports[`The test schema detects a invalid input as invalid 3`] = `
Processed {
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "Not one of (1, 2, 3, 4).",
      ],
      "value": 19,
    },
    "b": Processed {
      "nested": Object {
        "c": Processed {
          "errors": Array [
            "Not a valid integer.",
          ],
        },
      },
    },
    "d": Processed {
      "errors": Array [
        "Missing required field.",
      ],
    },
  },
}
`;

exports[`The test schema detects a invalid input as invalid 4`] = `
Processed {
  "nested": Object {
    "d": Processed {
      "errors": Array [
        "Missing required field.",
      ],
    },
  },
}
`;
